src/ucm/malloc/malloc_hook.c

```diff
@@ -0,0 +1,438 @@
+
+    ucs_spin_unlock(&ucm_malloc_hook_state.lock);
+}
+
+static int ucm_malloc_mmaped_ptr_remove_if_exists(void *ptr)
+{
+    unsigned i;
+
+    ucs_spin_lock(&ucm_malloc_hook_state.lock);
+    for (i = 0; i < ucm_malloc_hook_state.num_ptrs; ++i) {
+        if (ucm_malloc_hook_state.ptrs[i] == ptr) {
+            --ucm_malloc_hook_state.num_ptrs;
+            ucm_malloc_hook_state.ptrs[i] =
+                            ucm_malloc_hook_state.ptrs[ucm_malloc_hook_state.num_ptrs];
+            ucs_spin_unlock(&ucm_malloc_hook_state.lock);
+            return 1;
+        }
+    }
+    ucs_spin_unlock(&ucm_malloc_hook_state.lock);
+    return 0;
+}
+
+static int ucm_malloc_is_address_in_heap(void *ptr)
+{
+    int in_heap;
+
+    ucs_spin_lock(&ucm_malloc_hook_state.lock);
+    in_heap = (ptr >= ucm_malloc_hook_state.heap_start) &&
+              (ptr < ucm_malloc_hook_state.heap_end);
+    ucs_spin_unlock(&ucm_malloc_hook_state.lock);
+    return in_heap;
+}
+
+static int ucm_malloc_address_remove_if_mapped(void *ptr, const char *caller)
+{
+    int is_mapped;
+
+    if (ucm_malloc_is_address_in_heap(ptr)) {
+        is_mapped = 1;
+    } else {
+        is_mapped = ucm_malloc_mmaped_ptr_remove_if_exists(ptr);
+    }
+
+    ucm_trace("%s: %p is%s ours (heap [%p..%p])",  caller, ptr,
+              is_mapped ? "" : " not",
+              ucm_malloc_hook_state.heap_start, ucm_malloc_hook_state.heap_end);
+    return is_mapped;
+}
+
+static void ucm_malloc_allocated(void *ptr, const char *caller)
+{
+    if (ucm_malloc_is_address_in_heap(ptr)) {
+        ucm_trace("%s: %p in heap [%p..%p]", caller, ptr,
+                  ucm_malloc_hook_state.heap_start, ucm_malloc_hook_state.heap_end);
+    } else {
+        ucm_trace("%s: %p is mmapped", caller, ptr);
+        ucm_malloc_mmaped_ptr_add(ptr);
+    }
+}
+
+static void ucm_release_foreign_block(void *ptr)
+{
+#if !NVALGRIND
+    if (RUNNING_ON_VALGRIND) {
+        /* We want to keep valgrind happy and release foreign memory as well.
+         * Otherwise, it's safer to do nothing.
+         */
+        ucm_malloc_hook_state.free(ptr);
+    }
+#endif
+}
+
+void *ucm_malloc(size_t size, const void *caller)
+{
+    void *ptr;
+
+    ucm_malloc_hook_state.hook_called = 1;
+    if (ucm_global_config.alloc_alignment > 1) {
+        ptr = dlmemalign(ucm_global_config.alloc_alignment, size);
+    } else {
+        ptr = dlmalloc(size);
+    }
+    ucm_malloc_allocated(ptr, "malloc");
+    return ptr;
+}
+
+void *ucm_realloc(void *oldptr, size_t size, const void *caller)
+{
+    void *newptr;
+
+    ucm_malloc_hook_state.hook_called = 1;
+
+    if ((oldptr == NULL) || ucm_malloc_address_remove_if_mapped(oldptr, "realloc")) {
+        newptr = dlrealloc(oldptr, size);
+        goto out;
+    }
+
+    /* If pointer was created by original malloc(), allocate the new pointer
+     * with the new heap, and copy out the data. Then, release the old pointer.
+     */
+    newptr = dlmalloc(size);
```

Comment:
realloc does not guarantee that the pointer will not change, but if the pointer changes the data has to be copied.
However - now i see a bug here - i didn't handle a case when realloc shrinks the existing block.